plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -0,0 +1,426 @@
+        + ":" + coordinates.getArtifactId()
+        + ":" + coordinates.getVersion();
+  }
+
+  /**
+   * Returns asset key based on passed in {@link MavenPath} path value.
+   */
+  private String getAssetKey(final MavenPath mavenPath) {
+    // TODO: maybe sha1() the resulting string?
+    return mavenPath.getPath();
+  }
+
+  /**
+   * Finds component by key.
+   */
+  @Nullable
+  private OrientVertex findComponent(final StorageTx tx,
+                                     final Coordinates coordinates,
+                                     final OrientVertex bucket)
+  {
+    final String componentKeyName =
+        StorageFacet.P_ATTRIBUTES + "." + getRepository().getFormat().getValue() + "." + P_COMPONENT_KEY;
+    return tx.findComponentWithProperty(componentKeyName, getComponentKey(coordinates), bucket);
+  }
+
+  /**
+   * Returns a list of all component assets.
+   */
+  private List<OrientVertex> getComponentAssets(final OrientVertex component)
+  {
+    final List<Vertex> vertices = Lists
+        .newArrayList(component.getVertices(Direction.IN, StorageFacet.E_PART_OF_COMPONENT));
+    final List<OrientVertex> result = Lists.newArrayList();
+    for (Vertex v : vertices) {
+      if (v instanceof OrientVertex) {
+        result.add((OrientVertex) v);
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Selects a component asset by key.
+   */
+  @Nullable
+  private OrientVertex selectComponentAsset(final StorageTx tx,
+                                            final OrientVertex component,
+                                            final MavenPath artifactPath)
+  {
+    final String assetKey = getAssetKey(artifactPath);
+    final List<OrientVertex> assets = getComponentAssets(component);
+    for (OrientVertex v : assets) {
+      final NestedAttributesMap attributesMap = getFormatAttributes(tx, v);
+      if (assetKey.equals(attributesMap.get(P_ASSET_KEY, String.class))) {
+        return v;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Finds asset by key.
+   */
+  @Nullable
+  private OrientVertex findAsset(final StorageTx tx,
+                                 final MavenPath mavenPath,
+                                 final OrientVertex bucket)
+  {
+    final String assetKeyName =
+        StorageFacet.P_ATTRIBUTES + "." + getRepository().getFormat().getValue() + "." + P_ASSET_KEY;
+    return tx.findAssetWithProperty(assetKeyName, getAssetKey(mavenPath), bucket);
+  }
+
+  private NestedAttributesMap getAttributes(final StorageTx tx, final OrientVertex vertex) {
+    return tx.getAttributes(vertex);
+  }
+
+  private NestedAttributesMap getFormatAttributes(final StorageTx tx, final OrientVertex vertex) {
+    return getAttributes(tx, vertex).child(getRepository().getFormat().getValue());
+  }
+
+  /**
+   * Determines or confirms the content type for the content, or throws {@link InvalidContentException} if it cannot.
+   */
+  @Nonnull
+  private String determineContentType(final MavenPath mavenPath,
```

Comment:
Looks very similar to RawContentFacetImpl#determineContentType, just takes a different kind of path. Seems like a common utility method somewhere would be handy.